mail_send_error_activity/models/mail_mail.py

```diff
@@ -0,0 +1,62 @@
+# Copyright (C) Cetmix OU <http://www.cetmix.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import models
+
+
+class MailMail(models.Model):
+    _inherit = "mail.mail"
+
+    def _prepare_mail_error_activity_vals(self):
+        """
+        Prepare mail error activity values
+        :return: dict of values
+        """
+        user_ids = self.author_id.user_ids
+        if (
+            len(user_ids) == 0
+            or not (self.res_id and self.model)
+            or not hasattr(self.env[self.model], "activity_ids")
+        ):
+            return False
+        activity_type = self.env.ref(
+            "mail_send_error_activity.mail_activity_data_mail_send_error"
+        )
+        return {
+            "user_id": user_ids[0].id,
+            "activity_type_id": activity_type.id,
+            "summary": "{} {}".format(
+                activity_type.summary,
+                "".join([partner.email for partner in self.partner_ids]),
+            ),
+            "note": self.failure_reason,
+            "res_model_id": self.env["ir.model"]._get(self.model).id,
+            "res_id": self.res_id,
+        }
+
+    def _create_activity_on_mail_error(self):
+        """
+        Create an activity on mail send error
```

Comment:
Why is it necessary to put in a separate method of creating activiti?

Reply:
Resolved